mslice/util/mantid/_workspace_ops.py

```diff
@@ -0,0 +1,84 @@
+    "LessThan": operator.lt,
+    "GreaterThan": operator.gt,
+    "Or": operator.or_,
+    "And": operator.and_,
+    "Xor": operator.xor
+}
+
+def _binary_op(self, other, algorithm, result_info, inplace, reverse):
+    """
+    Delegate binary operations (+,-,*,/) performed on a workspace depending on type.
+
+    Unwrap mantid workspace (_raw_ws), perform operation, then rewrap.
+
+    Behaviour depends on type of other:
+    List - convert to numpy array, then pass to _binary_op_array method.
+    Numpy array - pass to _binary_op_array method.
+    Workspace wrapper (of the same type and dimensionality) - unwrap, apply operator to each bin, rewrap.
+    Else - try to apply operator to self._raw_ws. Works for numbers, unwrapped workspaces...
+
+    :param operator: binary operator (add, sub, mul, div)
+    :param other: object to add/sub/mul/div with self - can be list, numpy array, workspace, number...
+    :return: new workspace wrapper with same type as self.
+    """
+    if isinstance(other, list):
+        other = np.asarray(other)
+    if isinstance(other, self.__class__):
+        if _check_dimensions(self, other):
+            inner_res = _do_binary_operation(algorithm, self._raw_ws, other._raw_ws, result_info, inplace, reverse)
+        else:
+            raise RuntimeError("workspaces must have same dimensionality for binary operations (+, -, *, /)")
+    elif isinstance(other, np.ndarray):
+        inner_res = self._binary_op_array(_binary_operator_map[algorithm], other)
+    else:
+        inner_res = _do_binary_operation(algorithm, self._raw_ws, other, result_info, inplace, reverse)
+    return self.rewrap(inner_res)
+
+def _check_dimensions(self, workspace_to_check):
+    """check if a workspace has the same number of bins as self for each dimension"""
+    for i in range(self._raw_ws.getNumDims()):
+        if self._raw_ws.getDimension(i).getNBins() != workspace_to_check._raw_ws.getDimension(i).getNBins():
+            return False
+    return True
+
+def _attach_binary_operators():
+    def add_operator_func(attr, algorithm, inplace, reverse):
+        def op_wrapper(self, other):
+            result_info = lhs_info()
+            # Replace output workspace name with a unique temporary name hidden in ADS
+            if result_info[0] > 0:
+                result_info = (result_info[0], ('__MSLTMP' + str(uuid4())[:8],) + result_info[1][1:])
+            return _binary_op(self, other, algorithm, result_info, inplace, reverse)
+
+        op_wrapper.__name__ = attr
+        setattr(WorkspaceOperatorMixin, attr, op_wrapper)
+
+    operations = {
+        "Plus": ("__add__", "__radd__", "__iadd__"),
+        "Minus": ("__sub__", "__rsub__", "__isub__"),
+        "Multiply": ("__mul__", "__rmul__", "__imul__"),
+        "Divide": ("__truediv__", "__rtruediv__", "__itruediv__"),
+        "LessThan": "__lt__",
+        "GreaterThan": "__gt__",
+        "Or": "__or__",
+        "And": "__and__",
+        "Xor": "__xor__"
+    }
```

Comment:
Yes, both are needed. This dict maps the Mantid Algorithm name to the Python magic method name to be defined - both are strings. The `_binary_operator_map` maps the Mantid Algorithm name to an operator (callable) which is then used by the `_binary_op_array` method of `WorkspaceMixin` ( https://github.com/mantidproject/mslice/blob/master/mslice/workspace/workspace_mixin.py#L70 ) which in turn uses a method of `numpy_helper` ( https://github.com/mantidproject/mslice/blob/master/mslice/util/numpy_helper.py#L4 ). This method uses the callable to apply the actual operation.

The dict is really just used in the loop directly below it. I'd really rather have them close together for readability reason. I guess it could just be put in the loop as

```
for alg, attributes in {
        "Plus": ("__add__", "__radd__", "__iadd__"),
        "Minus": ("__sub__", "__rsub__", "__isub__"),
        "Multiply": ("__mul__", "__rmul__", "__imul__"),
        "Divide": ("__truediv__", "__rtruediv__", "__itruediv__"),
        "LessThan": "__lt__",
        "GreaterThan": "__gt__",
        "Or": "__or__",
        "And": "__and__",
        "Xor": "__xor__"
    }:
```

But I thought defining the dict first looks nicer.